tests/Aspire.Hosting.Containers.Tests/ContainerResourceBuilderTests.cs

```diff
@@ -104,6 +104,92 @@ public void WithImageSHA256ThrowsIfNoImageAnnotation()
         Assert.Equal("The resource 'testcontainer' does not have a container image specified. Use WithImage to specify the container image and tag.", exception.Message);
     }
 
+    [Theory]
+    [InlineData("redis", "redis", "latest", null)]
+    [InlineData("redis:latest", "redis", "latest", null)]
+    [InlineData("registry.io/library/rabbitmq", "registry.io/library/rabbitmq", "latest", null)]
+    [InlineData("postgres:tag", "postgres", "tag", null)]
+    [InlineData("kafka@sha256:01234567890abcdef01234567890abcdef01234567890abcdef01234567890ab", "kafka", null, "01234567890abcdef01234567890abcdef01234567890abcdef01234567890ab")]
+    [InlineData("registry.io/image:tag", "registry.io/image", "tag", null)]
+    [InlineData("host.com/path/to/image@sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "host.com/path/to/image", null, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")]
+    [InlineData("another.org/path/to/another/image:tag@sha256:9999999999999999999999999999999999999999999999999999999999999999", "another.org/path/to/another/image", null, "9999999999999999999999999999999999999999999999999999999999999999")]
+    public void WithImageMutatesContainerImageAnnotation(string reference, string expectedImage, string? expectedTag, string? expectedSha256)
+    {
+        using var builder = TestDistributedApplicationBuilder.Create();
+        var container = builder.AddResource(new TestContainerResource("testcontainer"));
+
+        container.WithImage(reference);
+
+        AssertImageComponents(container, null, expectedImage, expectedTag, expectedSha256);
+    }
+
+    [Fact]
+    public void WithImageThrowsWithConflictingTag()
+    {
+        using var builder = TestDistributedApplicationBuilder.Create();
+        var container = builder.AddResource(new TestContainerResource("testcontainer"));
+
+        Assert.Throws<ArgumentOutOfRangeException>(() => container.WithImage("image:tag", "anothertag"));
+    }
+
+    [Fact]
+    public void WithImageThrowsWithConflictingTagAndDigest()
+    {
+        using var builder = TestDistributedApplicationBuilder.Create();
+        var container = builder.AddResource(new TestContainerResource("testcontainer"));
+
+        Assert.Throws<ArgumentOutOfRangeException>(() => container.WithImage("image@sha246:abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd", "tag"));
+    }
+
+    [Fact]
```

Comment:
I don't like these tests - particularly the "Overwrite tag with latest if none is specified" part, but they represent the existing behaviour.

I would like to change this (see #5834), but it does introduce some possible breaking changes, so I've left it as is to let this PR narrowly focus on the first time user use case.

Reply:
At least we will have coverage when we decide to break it 😄